ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,464 @@
+module Ariadne.UI.Vty.Widget
+       ( WidgetNamePart(..)
+       , WidgetName
+       , WidgetEvent(..)
+       , WidgetEventResult(..)
+       , Widget(..)
+       , WidgetInitM
+       , WidgetDrawM
+       , WidgetEventM
+
+       , initWidget
+       , setWidgetState
+       , addWidgetChild
+       , setWidgetFocusList
+       , addWidgetEventHandler
+       , setWidgetDraw
+       , setWidgetDrawWithFocused
+       , setWidgetDrawWithFocus
+       , setWidgetScrollable
+       , setWidgetHandleEditKey
+       , setWidgetHandleKey
+       , setWidgetHandlePaste
+       , setWidgetHandleMouseDown
+       , setWidgetHandleScroll
+       , setWidgetHandleEvent
+
+       , getWidgetName
+       , getFocusRing
+       , findClosestFocus
+       , liftBrick
+       , widgetParentGetter
+       , widgetParentLens
+       , viewWidgetLens
+       , useWidgetLens
+       , assignWidgetLens
+       , widgetEvent
+
+       , drawWidget
+       , drawWidgetChild
+
+       , handleWidgetEditKey
+       , handleWidgetKey
+       , handleWidgetPaste
+       , handleWidgetMouseDown
+       , handleWidgetScroll
+       , handleWidgetEvent
+
+       -- Re-exports
+       , ReifiedLens'
+       , ReifiedLens(..)
+       ) where
+
+import Universum
+
+import Control.Lens (ReifiedLens', ReifiedLens(..), assign, makeLensesWith, to, (%=), (.=))
+import IiExtras (postfixLFields)
+
+import qualified Brick as B
+import qualified Brick.Focus as B
+import qualified Data.Map.Strict as Map
+
+import Ariadne.UI.Vty.Face
+import Ariadne.UI.Vty.Keyboard
+import Ariadne.UI.Vty.Scrolling
+
+----------------------------------------------------------------------------
+-- Types and instances
+----------------------------------------------------------------------------
+
+-- | Component of a unique widget name
+--
+-- Each widget has its child widgets indexed by this name component.
+-- Widget's unique name is constructed as a list of these components,
+-- corresponding to its location in widget tree.
+--
+-- Should be made into a union of widget-local name types.
+data WidgetNamePart
+  = WidgetNameSelf  -- Special name for adding widget itself to focus list
+
+  | WidgetNameMenu
+  | WidgetNameTree
+  | WidgetNameStatus
+  | WidgetNameHelp
+  | WidgetNameAbout
+  | WidgetNameLogs
+
+  | WidgetNameRepl
+  | WidgetNameReplInput
+
+  | WidgetNameAddWallet
+  | WidgetNameAddWalletNewName
+  | WidgetNameAddWalletNewPass
+  | WidgetNameAddWalletNewButton
+  | WidgetNameAddWalletRestoreName
+  | WidgetNameAddWalletRestoreMnemonic
+  | WidgetNameAddWalletRestorePass
+  | WidgetNameAddWalletRestoreFull
+  | WidgetNameAddWalletRestoreButton
+
+  | WidgetNameWallet
+  | WidgetNameWalletSendAddress
+  | WidgetNameWalletSendAmount
+  | WidgetNameWalletSendPass
+  | WidgetNameWalletSendButton
+
+  | WidgetNameAccount
+  deriving (Eq, Ord, Show)
+
+-- | Unique widget name, describing its location in widget tree
+type WidgetName = [WidgetNamePart]
```

Comment:
Probably would be better to make it `NonEmpty` (if I got it right that it must not be empty).